translations: Add Simplified Chinese language support

ioiotor opened the pull request:
Add Simplified Chinese language support

ducalex commented:
Thanks for your work! 92542 added bytes per app is a lot. I'd like to add Chinese support, but we'll have to either:

- Make Chinese support optional at build time (individual targets could opt into it, plus maybe an option in rg_tool.py)
- Or find a way to share fonts between apps. A shared data partition maybe.

> The game file names appear as garbled characters. 

My understanding is that we need to enable dynamic codepage and UTF-8 support in FATFS. However it says that it adds 480KB per app which is a lot...

<img width="932" height="860" alt="image" src="https://github.com/user-attachments/assets/2e1dbfb4-6377-47dc-a6d6-4fbd2e509b86" />

Commit:
translations: Add Simplified Chinese language support

ioiotor committed:
Font: Add Fusion Pixel 12px monospaced font
Added the fusion-pixel-12px-monospaced-zh_hans font and the converted font, which includes the first-level Chinese characters of the Chinese National Standard and the characters used in the Chinese translation.

ioiotor commented:
The font and the converted font files have been committed.  
They include the characters required for Chinese translation and a total of 3950 Chinese characters from the Level 1 set of the Chinese National Standard.

ioiotor committed:
Update translations.h

Modify [RG_LANG_DE] = "German" to [RG_LANG_DE] = "Deutsch" to resolve the merge conflict.

ioiotor committed:
Merge branch 'dev' into dev

ducalex commented:
Sorry for the long delay! I've tested the changes and it seems to work fine, though my ability to test is very limited. 

The big glyph table certainly slows down text drawing, glyph lookup will need to be optimized eventually, but for now it seems to still be perfectly usable.

I'd like to add flags to selectively enable Chinese support before merging the PR, may I commit to your branch? 

I'd likely add a flag that can be defined in the target's config.h to enable Chinese support, and maybe a flag for rg_tool.py like `--with-chinese-support` to force enable it for any target. Also maybe add code to force the Fusion font when Chinese language is selected in the menu.

ioiotor commented:
Thank you for taking the time to test. You are welcome to commit to my branch anytime. Additionally, I tested it on the esp32-p4, and the performance is much smoother.

ducalex committed:
rg_gui: Add special cases for Chinese support

When the language is set to Chinese, rg_gui_set_font will refuse to use a non-chinese font.

rg_gui_set_language_id will refuse to set language to Chinese if the font isn't present.

ducalex committed:
rg_gui: Set font to Chinese font when language is changed

ducalex commented:
One remaining issue is that in the language change dialog the Chinese option is all blocks because the other fonts do not have the glyphs for 简体中文.

Retro-Go doesn't currently have an easy way of drawing in different fonts so I'm not sure what to do with that... It's also not possible/difficult to add those glyphs to the other fonts (it can probably be done manually with some work).

Another could be that, when Chinese support is enabled, Retro-Go is built with **only** Fusion Pixel 12 and all the other fonts are removed?

csy19900206 commented:
Hello,
Regarding the issue where the "简体中文" option appears as blocks in the language selection dialog due to missing glyphs in the current fonts, I recommend the following approach for a clean and sustainable fix.
Proposed Solution: Replace the default font with Alibaba PuHuiTi (Ali Universal Font)
This font is freely licensed for commercial and open-source use, supports full Simplified Chinese glyphs, and maintains good readability in small sizes.

Font Preparation
Obtain the Alibaba PuHuiTi Regular .ttf file.  [alibabafonts](https://www.alibabafonts.com/#/home)
Character Subsetting
Include basic ASCII (for English UI elements) and a core set of Simplified Chinese characters.
Recommended subset:
ASCII range (32–126)
Common Chinese characters (e.g., 《通用规范汉字表》一级字库, ~3500 characters)
Necessary punctuation and symbols.

Remove other redundant fonts from the build to save space (if no longer needed).
Ensure the font loader is configured to use UTF‑8 encoding for Chinese text rendering.
